### PR TITLE
Add customer-managed encryption key (KMS) support to  GCS backend

### PR DESCRIPTION
**Pull request by SarahFrench**
# Description

Closes https://github.com/hashicorp/terraform/issues/24967

Currently the gcs backend allows users the option of supplying an encryption key from the client side, [customer-_supplied_ encryption keys](https://cloud.google.com/storage/docs/encryption/customer-supplied-keys), which is used as an additional layer of encryption in addition to Google-managed encryption keys. If a key isn't supplied in this way then Google-managed encryption keys are used in the GCS bucket by default. 

This PR adds the alternative option of using a customer-managed KMS key to encrypt the state files in the bucket (in addition to Google's own encryption). Details of the KMS key are included in Storage API calls, and retrieving the key is internal to the GCS service.

## Things to note

The new encryption method conflicts with the customer-_supplied_ encryption feature that already exists in the backend BUT there isn't a risk of data being lost due to this. When switching between the two types of encryption, in either direction, the running `terraform init -migrate-state` results in errors returned from the API and state being OK and accessible in the 'old' encryption style, still.

Users will need to remove the old method of encryption from the state file before beginning to use the other type of encryption. There needs to be an explicit write to state to remove the old encryption method, and after that the new encryption method can start to be used.

**I've updated documentation to give guidance about this, but I could use a review specifically about technical writing. This is the altered page in the deployment of this PR: https://terraform-8l6jf7ots-hashicorp.vercel.app/language/settings/backends/gcs**

Due to the conflict between different encryption methods, this PR changes the schema to mark these fields as conflicting. As a result of this, I've needed to stop the `encryption_key` field having a default value of `""`. This isn't a breaking change because the  (schema.ResourceData).Get function will supply the zero value for that field when it is not defined in the config (or ENV).

## Open questions

1. Can I improve my use of contexts in the new tests?

2. Do we need to add any code to help handle the transition between different configurations of the GCS backend?
- I've addressed this in documentation for now

3. ~~Technical writing review needed?  [See changes in deployment here](https://terraform-8l6jf7ots-hashicorp.vercel.app/language/settings/backends/gcs)~~ ✅

### CODE REVIEW

**Comment by SarahFrench** [comment]
I had another go implementing the new config fields as a nested block instead of a string and investigated the problem I had before:
- I can use the new backend ok in manual tests
- Automated tests fail to process the new schema containing a nested block and show this error :
>Unsupported attribute: An attribute named "kms_encryption_key" is not expected here.

<br>

I think I've tracked down the problem to this comment:

https://github.com/hashicorp/terraform/blob/a29fefbec78c1744106de66e4c277261b8ce4874/internal/configs/synth_body.go#L76-L77

And because of this I'm inclined to leave the new kms_encryption_key attribute as a string

File: internal/backend/remote-state/gcs/backend_test.go [hunk 1]
```diff
@@ -205,6 +256,119 @@ func setupBackend(t *testing.T, bucket, prefix, key string) backend.Backend {
...
+		key, err = c.CreateCryptoKey(ctx, reqCreateKey)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Logf("KMS key `%s` created successfully", key.Name)
+	}
+
+	// Get GCS Service account email, check has necessary permission on key
+	storageCtx := context.TODO()
+	sc, err := storage.NewClient(storageCtx, opts...) //reuse opts from KMS client
```
Comment (SarahFrench) [inline internal/backend/remote-state/gcs/backend_test.go:342]: Note: cannot reuse the backend's storage client (like in the `setupBucket` function) because the KMS key needs to exist before the backend buckets are made.

File: website/docs/language/settings/backends/gcs.mdx [hunk 2]
```diff
@@ -73,6 +73,28 @@ the path of the service account key. Terraform will use that key for authenticat
...
+
+### Customer-managed encryption keys (Cloud KMS)
+
+To get started, follow this guide: [Use customer-managed encryption keys](https://cloud.google.com/storage/docs/encryption/using-customer-managed-keys)
+
+If you want to remove customer-managed keys from your backend configuration, or change to a different customer-managed key, Terraform _can_ manage a state migration without manual intervention. This is because customer-managed encryption keys are stored in GCP and are accessible during the state migration process. However, these changes do not fully come into effect until the first write operation is performed on the state file after state migration occurs. In the first write operation after state migration the file will be decrypted with the old key and then written with the new encryption method. This is equivalent to the [rewrite](https://cloud.google.com/storage/docs/gsutil/commands/rewrite) operation described in the customer-supplied encryption keys section above. Because of the importance of the first write to state after state migration, you should not delete old KMS keys until any state file(s) encrypted with that key have been updated.
```
Comment (mgarrell777) [inline website/docs/language/settings/backends/gcs.mdx:92]:
```suggestion
If you want to remove customer-managed keys from your backend configuration or change to a different customer-managed key, Terraform _can_ manage a state migration without manual intervention. This ability is because GCP stores customer-managed encryption keys and are accessible during the state migration process. However, these changes do not fully come into effect until the first write operation occurs on the state file after state migration occurs. In the first write operation after state migration, the file decrypts with the old key and then writes with the new encryption method. This method is equivalent to the [rewrite](https://cloud.google.com/storage/docs/gsutil/commands/rewrite) operation described in the customer-supplied encryption keys section. Because of the importance of the first write to state after state migration, you should not delete old KMS keys until any state file(s) encrypted with that key update.
```

Suggest adding "ability" after "This" to avoid a "This" with no subject. Suggest changing "are stored in GCP" to "GCP stores" to avoid a passive voice style issue. Suggest changing "will be decrypted" and "is performed" to "decrypts" and "occurs" to avoid future tense and passive voice style issues. Suggest changing "have been updated" to "update" to avoid a passive voice style issue.

File: website/docs/language/settings/backends/gcs.mdx [hunk 3]
```diff
@@ -73,6 +73,28 @@ the path of the service account key. Terraform will use that key for authenticat
...
+
+Customer-managed keys do not need to be sent in requests to read files from GCS buckets, as decryption is handled automatically within GCS. This means that if you use the `terraform_remote_state` [data source](https://www.terraform.io/language/state/remote-state-data) to access KMS-encrypted state you do not need to specify the KMS key in the data source's `config` object.
```
Comment (mgarrell777) [inline website/docs/language/settings/backends/gcs.mdx:94]:
```suggestion
Customer-managed keys do not need to be sent in requests to read files from GCS buckets because decryption occurs automatically within GCS. This process means that if you use the `terraform_remote_state` [data source](https://www.terraform.io/language/state/remote-state-data) to access KMS-encrypted state, you do not need to specify the KMS key in the data source's `config` object. 
```

Suggest changing "is handled" to "occurs" to avoid a passive voice style issue. Suggest adding "process" after "This" to avoid a "This" without a subject.

File: website/docs/language/settings/backends/gcs.mdx [hunk 4]
```diff
@@ -73,6 +73,28 @@ the path of the service account key. Terraform will use that key for authenticat
...
+
+Customer-managed keys do not need to be sent in requests to read files from GCS buckets, as decryption is handled automatically within GCS. This means that if you use the `terraform_remote_state` [data source](https://www.terraform.io/language/state/remote-state-data) to access KMS-encrypted state you do not need to specify the KMS key in the data source's `config` object. 
+
+~> **Important:** To use customer-managed encryption keys, you will need to create a key and give your project's GCS service agent permission to use it with the Cloud KMS CryptoKey Encrypter/Decrypter predefined role.
```
Comment (mgarrell777) [inline website/docs/language/settings/backends/gcs.mdx:96]:
```suggestion
~> **Important:** To use customer-managed encryption keys, you need to create a key and give your project's GCS service agent permission to use it with the Cloud KMS CryptoKey Encrypter/Decrypter predefined role. 
```

Suggest changing "will need" to "need" to avoid a future tense style issue.

File: website/docs/language/settings/backends/gcs.mdx [hunk 5]
```diff
@@ -73,6 +73,28 @@ the path of the service account key. Terraform will use that key for authenticat
...
+
+Customer-managed keys do not need to be sent in requests to read files from GCS buckets, as decryption is handled automatically within GCS. This means that if you use the `terraform_remote_state` [data source](https://www.terraform.io/language/state/remote-state-data) to access KMS-encrypted state you do not need to specify the KMS key in the data source's `config` object. 
+
+~> **Important:** To use customer-managed encryption keys, you will need to create a key and give your project's GCS service agent permission to use it with the Cloud KMS CryptoKey Encrypter/Decrypter predefined role. 
+
 ## Configuration Variables
 
 !> **Warning:**  We recommend using environment variables to supply credentials and other sensitive data. If you use `-backend-config` or hardcode these values directly in your configuration, Terraform will include these values in both the `.terraform` subdirectory and in plan files. Refer to [Credentials and Sensitive Data](/language/settings/backends/configuration#credentials-and-sensitive-data) for details.
```
Comment (mgarrell777) [inline website/docs/language/settings/backends/gcs.mdx:100]:
```suggestion
!> **Warning:**  We recommend using environment variables to supply credentials and other sensitive data. If you use `-backend-config` or hardcode these values directly in your configuration, Terraform includes these values in both the `.terraform` subdirectory and in plan files. Refer to [Credentials and Sensitive Data](/language/settings/backends/configuration#credentials-and-sensitive-data) for details.
```

Suggest changing "will include" to "includes" to avoid a future tense style issue.

File: website/docs/language/settings/backends/gcs.mdx [hunk 6]
```diff
@@ -86,7 +108,7 @@ The following configuration options are supported:
   (Optional) Local path to Google Cloud Platform account credentials in JSON
   format.  If unset, [Google Application Default
   Credentials](https://developers.google.com/identity/protocols/application-default-credentials)
-  are used.  The provided credentials must have Storage Object Admin role on the bucket.
+  are used.  The provided credentials must have the Storage Object Admin role on the bucket.
```
Comment (mgarrell777) [inline website/docs/language/settings/backends/gcs.mdx:111]: I can't include a direct suggest change across a deletion. So I suggest rewording the sentence to "If unset, the path uses [Google Application Default Credentials]..." to avoid a passive voice style issue.

**Review by mgarrell777 — APPROVED** [review]
@SarahFrench Thanks very much for your contribution with this PR. I have some suggestions for wording changes, but overall things look good to me.

File: website/docs/language/settings/backends/gcs.mdx (hunk 6 above)
Comment (SarahFrench) [reply website/docs/language/settings/backends/gcs.mdx:111]: Will do, thanks!

**Comment by SarahFrench** [comment]
Note: just force pushed to squash a few commits and clean up commit messages

@mgarrell777 - Thanks for all your help, I'll watch out for using passive voice in future!

File: internal/backend/remote-state/gcs/backend.go [hunk 7]
```diff
@@ -212,5 +223,15 @@ func (b *Backend) configure(ctx context.Context) error {
 		b.encryptionKey = k
 	}
 
+	// Customer-managed encryption
+	kmsName := data.Get("kms_encryption_key").(string)
+	if kmsName != "" {
+		kc, err := backend.ReadPathOrContents(kmsName)
```
Comment (alisdair) [inline internal/backend/remote-state/gcs/backend.go:229]: I'm confused by the use of `ReadPathOrContents`, both here and when processing `encryption_key`. From reading the docs I understood `encryption_key` to be an inline base64 encoded key, and `kms_encryption_key` to be a GCP path reference.

`ReadPathOrContents` seems to try to read a local file by treating its argument as a relative path, and only upon failing returns the path. If that functionality is desirable, I think we should document it; if it's not, I think we can just skip calling this function altogether.

File: internal/backend/remote-state/gcs/backend_test.go [hunk 8]
```diff
@@ -205,6 +255,119 @@ func setupBackend(t *testing.T, bucket, prefix, key string) backend.Backend {
...
+		}
+		key, err = c.CreateCryptoKey(ctx, reqCreateKey)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Logf("KMS key `%s` created successfully", key.Name)
+	}
+
+	// Get GCS Service account email, check has necessary permission on key
+	storageCtx := context.TODO()
```
Comment (alisdair) [inline internal/backend/remote-state/gcs/backend_test.go:340]: This has no practical impact and is just a style suggestion, but I think it's appropriate to use `context.Background` here, because we don't seem to intend to update this with any other parent context in the future.

File: internal/backend/remote-state/gcs/backend_test.go (hunk 1 above)
Comment (alisdair) [reply internal/backend/remote-state/gcs/backend_test.go:341]: This is a useful note and might be worth inlining as a code comment, too.

**Review by alisdair — APPROVED** [review]
These changes look good to me!

I like the choice to make the two encryption options conflict rather than attempt to support automatic migration between the two. That kind of change ought to be rare and I think it's better to ask our users to carefully handle it manually than take responsibility for doing it ourselves.

File: internal/backend/remote-state/gcs/backend.go (hunk 7 above)
Comment (SarahFrench) [reply internal/backend/remote-state/gcs/backend.go:229]: Yeah I was similarly confused- I chose to use that function to be consistent with `encryption_key`. I'm not keen to remove it from `encryption_key` because it could be a breaking change for some. I'll add it to the docs for both fields!

Comment (apparentlymart) [reply internal/backend/remote-state/gcs/backend.go:229]: This `ReadPathOrContents` helper is a very old shim for when we decide to standardize on passing settings around in strings rather than using filenames in most cases, so folks would be able to construct the input in various different ways rather than always forcing a file on disk.

However as well as now being very legacy, I'm also skeptical about it being used in a backend configuration context because dynamic expressions aren't allowed here anyway... anything assigned to a backend configuration must always be static.

If we do want to support both literal assignment and assignment by filename then for modern designs I would suggest separate arguments for each and a validation rule preventing using them together. Overloading a single argument to support two different usages was a pragmatic compatibility shim for some existing arguments that already had the name we wanted to keep, but for new designs we can use an argument whose name ends in `_file` to accept the filename version separately from the unsuffixed direct value version.

Agreed on not making a breaking change to the existing argument right now, but I suggest following the new pattern for the new argument, at the expense of consistency, and planning to break this _eventually_ to make the older argument consistent with the new design at a later date.

Comment (SarahFrench) [reply internal/backend/remote-state/gcs/backend.go:229]: That's useful background info about it being a shim, in that case I'll avoid using it for the new field like you suggested. 

If we eventually break the `encryption_key` field and add `encryption_key_file` (or equiv.) I'll avoid updating the documentation in this PR;  we probably don't want to encourage more people to creating configurations that would experience the breaking change!

**Comment by SarahFrench** [comment]
Thanks for your help @alisdair ! I'm feeling a lot more confident about getting this merged. And also I'm glad that you agree that we should leave changing between encryption methods as a manual steps for users, instead of automating it 😅 

What should the next steps for this PR be? Is it ok for me to merge, and should I update CHANGELOG.md afterwards? Or should I [leave that](https://github.com/hashicorp/terraform/blob/main/.github/CONTRIBUTING.md#pull-request-lifecycle) to the core team? Here's a proposed enhancement entry:

```
* backend/gcs: Add `kms_encryption_key` argument, to allow encryption of state files using Cloud KMS keys
```

**Comment by alisdair** [comment]
Yep, it's fine for you to merge, and please do update the CHANGELOG after doing so. That looks good to me—house style is to end the sentence with a period, and don't forget the GitHub issue at the end (i.e. `[GH-31786]`).
